### ISSUE TITLE
Object without `properties` throws

### ISSUE THREAD

**Issue by chmac** [mozilla-services/react-jsonschema-form#44]
Reproduction from the example:

``` js
const schema = {
  title: "Todo Tasks",
  type: "object",
  required: ["title"],
  properties: {
    title: {type: "object", title: "Title"},
    done: {type: "boolean", title: "Done?", default: false}
  }
};
```

The `title` field has been changed to `type: "object"`. This causes the package to throw unless there is a `properties` key on the `title` field. In my (admittedly very light) testing, the schema is valid without the `properties` keyword, but I could be entirely wrong about that.

The error I see is `Uncaught TypeError: Cannot convert undefined or null to object`. Adding `properties: {}` solves the issue for me.

Not sure if this is really a bug or not. Although I guess there would be no harm in checking for the existence of `properties` rather than assuming it exists. I'll see if I can find time to come up with a PR on that front if it would be helpful.

**Comment by n1k0** [comment]
> the schema is valid without the properties keyword, but I could be entirely wrong about that.

Confirmed, an object field doesn't required a properties parameter.

> I'll see if I can find time to come up with a PR on that front if it would be helpful.

That'd be just plain great, thank you :)

**Comment by chmac** [comment]
Done, without any testing at all...

**Comment by chmac** [comment]
@n1k0 Trying to get the tests running locally like `npm install --dev` but having some challenges. First time my dev VM ran out of memory. I killed some other big memory hogs and started again. Now it's been running for a while, maybe 15 minutes, and is complaining about missing repositories, and permissions errors in `~/.npm`. Any suggestions?

**Comment by n1k0** [comment]
I suggest to use [nvm](https://github.com/creationix/nvm) to install a separated nodejs environment and deal with different versions much more easily.

After installing it, you'll have node versions it managed installed in your homedir, which usually prevents permission errors. I suggest you `nvm install 4 && nvm use 4` to start hacking on this specific project. Don't forget to `rm -rf ./node_modules` and run `npm install` again when you're done.

This usually works everytime :)

**Comment by n1k0** [comment]
I'm realizing we can't really support `object` fields without any properties described, as that would mean an empty form or fieldset. We should probably raise instead, or at least trigger a warning in the console to notify the developer about that.

**Comment by chmac** [comment]
If the field is optional, and has `additionalProperties: true` then it could simply be skipped from the form. But I think the more useful strategy would be to allow a custom widget for fields of type object. Then we could implement file upload widgets and other whacky stuff which produces objects.

**Comment by n1k0** [comment]
> But I think the more useful strategy would be to allow a custom widget for fields of type object

This is already achievable by providing a [custom `SchemaField`](https://github.com/mozilla-services/react-jsonschema-form#custom-schemafield). What else would be missing? Documentation perhaps...

**Comment by chmac** [comment]
I think we looked at that, but the custom SchemaField is applied to every field in the form I think. Is that right? I'd suggest that supporting the `ui:widget` on object fields would be nice. It would make it easy for me to provide a custom component for just one field.

The use case for us is uploading images to services like cloudinary, where we get back an object, and we want to save it as part of the form.

It took me a while to figure out that the object field doesn't support the `ui:widget` option. But maybe I misunderstood the intention of that option, because I also didn't realise that the label is not rendered as part of the widget.

Bottom line, we ended up building custom forms because we couldn't get the pieces to work right, and didn't have enough time to figure out the package. We're looking for a general form package for react and this package looks like it has the best potential, and the UI schema concept seems like the way to go. If we could easily override the complete rendering of object fields, and there was an easy way to apply semantic UI classes across the form, then we'd have a very solid, and flexible form manager...

**Comment by n1k0** [comment]
Thanks for the valuable feedback. Basically in the formbuilder we used a custom SchemaField to wrap a regular SchemaField: https://github.com/Kinto/formbuilder/blob/master/formbuilder/components/EditableField.js#L189-L200

I understand this isn't the easiest possible way, but that's still powerful. For handling custom object field, that would give something like this (pseudo code):

```jsx
import SchemaField from "react-jsonschema-field/lib/components/fields/SchemaField"

class MyObjectField extends React.Component {
  constructor(props) {
    super(props);
  }

  render() {
    return <div>Custom object field</div>;
  }
}

const MySchemaField = ({schema, ...props}) => {
  if (schema.type === "object") {
    return <MyObjectField {...props} />;
  }
  return <SchemaField {...props} />
} 

render(<Form SchemaField=MySchemaField/>, ...);
```

I'll try thinking if we could easily provide a more direct way of overriding an object field though.

**Comment by chmac** [comment]
That would override **all** object fields I think, right? We'd need to effectively right our own switch statement, based on the name of the field and it's position in the schema and so on, which is where I see the potential for error. Also, I like the idea of building a generic component which we can use in multiple forms. Modifying the UI schema feels like the correct way to specify how a specific field should be rendered, rather than building custom logic to handle that.

Once I eventually get nvm and the tests running, I'd thought to submit a PR that adds `ui:widget` support to object fields. It should be reasonably simple to copy the functionality from one of the other fields was my thinking...

**Comment by n1k0** [comment]
I agree that having support for custom object and array fields through `ui:widget` is the way to go.

**Comment by chmac** [comment]
@n1k0 I like the warning idea. My thinking here is that we might use JSON schemas which we want to partially render into forms. The JSON schema might also be used to validate data going into the database. If we do that, we potentially allow arbitrary objects like `type: 'object', additionalProperties: true`, and apparently a JSON schema is valid even with the `properties:` key on a type `object`. For that reason, I'd advocate for the logging option rather than throwing.

However, now that you've solved our use case in #77, I'm happy either way. :-)

**Comment by epicfaace** [comment]
@chmac's example seems to be working in the latest version here (https://jsfiddle.net/epicfaace/4xznjm2k/), so I'm closing this.
